### ISSUE TITLE
pext() in AllTrue()/AllFalse() on AVX2::Mask<U16>

### ISSUE THREAD

**Issue by yzazik** [google/highway#436]
Hello,
Straight to the point.
Why `AllTrue()` and `AllFalse()` need `_pext_u64()` for `Mask<U16>` but not for `Mask256<U16>`?
[https://godbolt.org/z/q7bc1cGaa](https://godbolt.org/z/q7bc1cGaa)
All other types (U8,U32,U64) have no pext.
Thank you.

**Comment by jan-wassenberg** [comment]
Hi @yzazik , this is because AVX2 provides only _mm256_movemask_epi8 and _mm256_movemask_ps (also usable for U32), but not _mm256_movemask_epi16.
Do you have a suggestion for better codegen? It is true that PEXT is slow on (all but very recent) AMD because it is microcoded.

**Comment by yzazik** [comment]
Well, if I play nice and use only **valid** `Mask`s (all lanes have either all-ones or all-zeroes), `_mm256_movemask_epi8()` does not check every bit of U8 for `AllFalse(Mask<U8>)` but only every 8th, then for `AllFalse(Mask<U16>)` would check every 8th too (although I'm interested in every 16th) and it would be enough too and I don't need expensive `pext` to tell me more (it might tell me more since it's part of `bitsFromMask()` but I don't care, I already got what I need from `_mm256_movemask_epi8()` for AllFale()).
In other words, what's my reward for playing nice, and `pext` definitely kills the (performance) mood, why not just remove it (not for `bitsFromMask(SizeTag<2>)`, but just use `bitsFromMask(SizeTag<1>)` instead).
But another thing I don't get is why on 256-bit native arch would I want a different code gen in `AllFalse()/AllTrue()` for `Mask<U16>` (an implicit alias for `Mask256<U16>`) and `Mask256<U16>` (specified explicitly), that is rather unexpected.

**Comment by jan-wassenberg** [comment]
Ah, I see, we can indeed avoid the pext here. I will send a patch shortly, thanks for suggesting this!

For the second issue, Mask<Simd<uint16, ?>> is fine but Mask256 is an internal-only type, and if used, the template arg should be <uint16_t>, not Simd<uint16_t, ?>.

**Comment by yzazik** [comment]
Sure, I meant Simd<>, it was just less typing :) Thanks for the quick response, much appreciate it.

**Comment by jan-wassenberg** [comment]
:) You're welcome!
